File: Himall.Web/Areas/Web/Scripts/ProductComment.js
﻿var isShowComment = false;
//商品评价
function GetProductComment() {
    if (!isShowComment) {
        var idList = [],// id存放列表
            uuid = 1,// 用来统计请求次数
            pid = $('#gid').val(),// 商品id
            getData = function (pageno, commenttype) {
                $.ajax({
                    type: 'get',
                    url: '/Product/GetCommentByProduct?pId=' + pid + '&pageNo=' + pageno + '&pageSize=' + 10 + '&commentType=' + commenttype,
                    dataType: 'html',
                    cache: true,// 开启ajax缓存
                    success: function (data) {
                        if (data) {
                            var data = (new Function('return ' + data))();
                            render(data, commenttype);
                        }
                    },
                    error: function (e) {
                        //
                    }
                });
            },
            template = [
                '<div id="comment-0" class="mc">',
                '<div class="item clearfix">',
                '<div class="user">',
                '<div class="u-icon">',
                '<a target="_blank" title="查看TA的全部评价">',
                '<div class="u-name">',
                '<span class="u-level">',
                '<span class="u-address">',
                '<div class="i-item">',
                '<div class="o-topic">',
                '<span class="date-comment">',
                '<div class="comment-content clearfix">'
            ],// 模板数据
            render = function (data, index) {
                var dom = $('#comments-list'),
                    str = '',
                    i,
                    e,
                    con = data.comments;
                for (i = 0; e = con[i++];) {
                    str += template[1]
                            + template[2]
                        + template[3]
                        + '<img src="' + ((e.Picture == null || e.Picture == "") ? "/Areas/Web/images/avatar.png" : e.Picture) + '">'
                                    + e.UserName
                                    + '</div></div>'
                                + template[8]
                                	+'<span class="star sa' + e.ReviewMark + '"></span>'
                                    + template[11]
                        + '<dl class="clearfix"><dt class="reviewdate">发表于' + template[10] + e.ReviewDate.substr(0, 10) + '</span></dt><dd>' + e.ReviewContent + '<div class="after-service-img clearfix">';

                    for (var j = 0; j < e.Images.length; j++) {
                        str += '<img height="40" src="' + e.Images[j].CommentImage + '"/><span class="glyphicon glyphicon-zoom-in"></span>'
                    }
                    str += '</div>';
                    str += '<div class="preview-img"><img src="" /></div>'
                    if (e.ReplyContent != "" && e.ReplyContent != "暂无回复" && e.ReplyContent != null) {
                        str += '<div class="shop-reply">商家回复：' + e.ReplyContent + '</div>';//e.ReplyDate
                    }

                    str += '</dd><div class="comment-sku">';
                    if (e.Color != '')
                        str += '<p>' + e.ColorAlias + '：' + e.Color + '</p>';
                    if (e.Size != '')
                        str += '<p>' + e.SizeAlias + '：' + e.Size + '</p>';
                    if (e.Version != '')
                        str += '<p>' + e.VersionAlias + '：' + e.Version + '</p>';
                    str += '</div></dl>';

                    if (e.AppendDate != null && e.AppendDate != "") {
                        str += '<dl class="clearfix"><dt>收货' + GetDateDiff(e.FinishDate, e.AppendDate) + '后追加评价：</dt><dd>' + e.AppendContent + '<div class="after-service-img clearfix">';
                        for (var k = 0; k < e.AppendImages.length; k++) {
                            str += '<img  height="40" src="' + e.AppendImages[k].CommentImage + '"/>'
                        }

                        str += '</div>';
                        str += '<div class="preview-img"><img src="" /></div>'
                        if (e.ReplyDate != null && e.ReplyAppendContent != null && e.ReplyAppendContent != "" && e.ReplyAppendContent != "暂无回复") {
                            str += '<div class="shop-reply">商家回复：' + e.ReplyAppendContent + '</div>';//e.ReplyDate
                        }

                        str += '</dd></dl>';
                    }

                    str += '</div></div></div>';
                }

                var commentsItem = $('#comment-' + index, dom);
                if (commentsItem.length > 0) {
                    var pagin = $('.pagin', commentsItem);
                    commentsItem.children(':not(.pagin)').remove();
                    pagin.before(str);
                }
                else {
                    var pagin = commonJS.generatePaging(1, data.totalPage, function (pi) {
                        getData(pi, index);
                    });
                    commentsItem = $('<div id="comment-{0}" class="mc">{1}</div>'.format(index, str));
                    commentsItem.append(pagin.div);
                    dom.append(commentsItem);
                }

                // 只会执行一次 生成评论条数
                if (uuid++ === 1) {
                    var total = data.goodComment + data.badComment + data.comment,
                        g = data.goodComment,
                        c = data.comment,
                        b = data.badComment,
                        hasAppend = data.hasAppend,
                        hasImages = data.hasImages,
                        e = total == 0 ? 100 : Math.round((g / total).toFixed(2) * 100),// 取整 误差为1
                        f = total == 0 ? (e == 100 ? 0 : 100) : Math.round((c / total).toFixed(2) * 100),
                        j = (100 - e - f) < 0 ? 0 : Math.round(100 - e - f),//防止少于0
                        arr = [total, g, c, b, hasImages, hasAppend],
                        arr1 = [e, f, j];
                    $('#id_comment_btn').find('li').each(function (i, e) {
                        $(e).find('a').append('<em>(' + arr[i] + ')</em>');
                    });
                    $('#i-comment strong').empty().prepend(e + '%');
                    $('#praiseRate').empty().prepend(e + '%');
                    $('#i-comment .percent').find('span').each(function (i, e) {
                        $(e).html('(' + arr1[i] + '%)');
                    });
                    $('#i-comment .percent').find('div').each(function (i, e) {
                        $(e).css({ width: arr1[i] + 'px' });
                    });
                }
            };

        getData(1, 0);//第一次数据渲染 参数：1:第1页 0:所有评论

        $('#id_comment_btn').find('li').each(function (i, e) {
            $(e).bind('click', function () {
                var dom = $(this);
                dom.siblings('.active').removeClass('active');
                dom.addClass('active');
                if ($('#comment-' + i).length > 0) {
                    $('#comments-list .mc').each(function () {
                        $(this).hide();
                    });
                    $('#comment-' + i).show();
                } else {
                    $('#comments-list .mc').each(function () {
                        $(this).hide();
                    });
                    getData(1, i);// 按需载入评论数据 数据只生成一次 i代表类型
                }
                return false;// 阻止冒泡
            });
        });
        isShowComment = true
    }
}

function GetDateDiff(startDate, endDate) {
    var daysRound = GetDateDiffResult(startDate, endDate, "day");
    var hoursRound = GetDateDiffResult(startDate, endDate, "hour");
    var minutesRound = GetDateDiffResult(startDate, endDate, "minute");

    if (hoursRound > 24) {
        return parseInt(daysRound) + "天";
    } else if (minutesRound > 60) {
        return parseInt(hoursRound) + "小时";
    } else {
        if (parseInt(minutesRound) <= 0) {
            return "1分钟";
        }
        return parseInt(minutesRound) + "分钟";
    }
}


function GetDateDiffResult(startTime, endTime, diffType) {
    startTime = startTime.replace(/\-/g, "/");
    endTime = endTime.replace(/\-/g, "/");
    diffType = diffType.toLowerCase();
    var sTime = new Date(startTime); //开始时间
    var eTime = new Date(endTime); //结束时间
    var timeType = 1;
    switch (diffType) {
        case "second":
            timeType = 1000;
            break;
        case "minute":
            timeType = 1000 * 60;
            break;
        case "hour":
            timeType = 1000 * 3600;
            break;
        case "day":
            timeType = 1000 * 3600 * 24;
            break;
        default:
            break;
    }
    return parseInt((eTime.getTime() - sTime.getTime()) / parseInt(timeType));
}

File: Himall.DTO/Product/ProductComment.cs
using Himall.Entities;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Himall.DTO
{
	public class ProductComment
	{
		public long Id { get; set; }
		public long ProductId { get; set; }
		public long ShopId { get; set; }
		public string ShopName { get; set; }
		public long UserId { get; set; }
		public string UserName { get; set; }
		public string Email { get; set; }
		public string ReviewContent { get; set; }
		public System.DateTime ReviewDate { get; set; }
		public string ReplyContent { get; set; }
		public Nullable<System.DateTime> ReplyDate { get; set; }
		public int ReviewMark { get; set; }
		public long SubOrderId { get; set; }
		public string AppendContent { get; set; }
		public Nullable<System.DateTime> AppendDate { get; set; }
		public string ReplyAppendContent { get; set; }
		public Nullable<System.DateTime> ReplyAppendDate { get; set; }
		public Nullable<bool> IsHidden { get; set; }
        /// <summary>
        /// 商品名称
        /// </summary>
        public string ProductName { get; set; }
        /// <summary>
        /// 规格编号
        /// </summary>
        public string SkuId { get; set; }
        /// <summary>
        /// 规格内容
        /// </summary>
        public string SKU { get; set; }
        public List<ProductCommentImage> Images { get; set; }
	}

	public class ProductCommentImage
	{
		public long Id { get; set; }
		public string CommentImage { get; set; }
		public long CommentId { get; set; }
		public int CommentType { get; set; }
	}

	public class ProductCommentModelDTO
	{
		/// <summary>
		/// 头像
		/// </summary>
		public string Picture { get; set; }
		public string UserName { get; set; }
		public string ReviewContent { get; set; }
		public string ReviewDate { get; set; }
		public string ReplyContent { get; set; }
		public string ReplyDate { get; set; }
		public int ReviewMark { get; set; }
		public string AppendContent { get; set; }
		public string AppendDate { get; set; }
		public string ReplyAppendContent { get; set; }
		public string ReplyAppendDate { get; set; }

		/// <summary>
		/// 购买时间
		/// </summary>
		public string BuyDate { get; set; }

		/// <summary>
		/// 订单完成时间
		/// </summary>
		public string FinishDate { get; set; }

		public string Color { get; set; }
		public string Version { get; set; }
		public string Size { get; set; }
		public string ColorAlias { get; set; }
		public string SizeAlias { get; set; }
		public string VersionAlias { get; set; }

		public string Sku { 
			get {
				List<string> arrstr = new List<string>();
				if (!string.IsNullOrEmpty(this.Color))
					arrstr.Add(" " + this.ColorAlias + "：" + this.Color);
				if (!string.IsNullOrEmpty(this.Size))
					arrstr.Add(" " + this.SizeAlias + "：" + this.Size);
				if (!string.IsNullOrEmpty(this.Version))
					arrstr.Add(" " + this.VersionAlias + "：" + this.Version);

				if (arrstr.Count > 0)
					return string.Join(";", arrstr.ToArray());
				else
					return string.Empty;
			}
		}

		public List<ProductCommentImageInfo> Images { get; set; }
		public List<ProductCommentImageInfo> AppendImages { get; set; }
	}
}
